modules/quality_modules.nf

```diff
@@ -3,23 +3,43 @@ process assessSimpleQuality {
     tag { sample }
 
     input:
-    tuple val(sample), path(fasta), path(filteredbam)
+    tuple val(sample), path(fasta), path(filteredbam), path(compositebam), path(compositebambai), path(krakenreport)
 
     output:
     path ("${sample}_quality.csv")
 
+    // Script to put all wanted sample values together
     script:
+    // If running kraken2, setup to calculate % viral
+    if ( params.kraken_db ) {
+        checkKraken     = true
+        krakenHeader    = ",percent_viral_reads"
+    } else {
+        checkKraken     = false
+        krakenHeader    = ""
+    }
     """
+    # For kraken2 if it is available/input
+    if $checkKraken; then
+        KRAKEN_IDS=`{ grep -P "\\s+${params.kraken_viral_id}\\s+" $krakenreport | cut -d\$'\t' -f 2 || :; }`
+        ALL_READS=`samtools view -c $compositebam`
+        VIRAL_PERCENT=,`awk -v id_count="\$KRAKEN_IDS" -v total_reads="\$ALL_READS" 'BEGIN {if (id_count=="") {print 0} else {print (id_count*2)/total_reads * 100}}'`
+    else
+        VIRAL_PERCENT=""
+    fi
+
     # Header
-    printf "sample,num_reads_mapped,avg_read_depth,total_n_count,proportion_basecalled\n" > ${sample}_quality.csv
+    printf "sample,num_reads_mapped,mean_sequencing_depth,median_sequencing_depth,num_consensus_n,genome_completeness${krakenHeader}\n" > ${sample}_quality.csv
 
     # Stats we want
     NUM_READS=`samtools view -F 0x04 -c $filteredbam`
-    AVG_COV=`samtools depth -a $filteredbam | awk '{c++;s+=\$3} END {if (c=='0') {print "NA"} else {print s/c}}'`
-    N_COUNT_PLUS_PROP=`seqtk comp $fasta | awk '{ OFS = "," }{ x=\$3+\$4+\$5+\$6;y=\$2; if (y=='0') {print "NA,NA"} else print y-x,1-((y-x)/y) }'`
+    AVG_PLUS_MEDIAN_COV=`samtools depth -a $filteredbam | \
```

Comment:
Not as important for the overall flow of the pipeline in standard cases, but do you know what happens if there's an error in some of these commands (e.g., if `$filteredbam` was an invalid BAM file)? Would the nextflow pipeline crash, or would `AVG_PLUS_MEDIAN_COV` report an invalid value (e.g,. an empty string)?

Reply:
I believe I've tested it on empty files and its been good. Invalid files I don't know. I think that would be a good idea to add to the test data!

Also thanks for the kraken2 info. You are right that was a good addition

Reply:
Thanks. I think tests with empty files is probably good enough. I cannot see when you would encounter corrupted files except in cases where the hard drive filled up before a BAM file got written, but then Nextflow itself should crash in that scenario.

Reply:
And what you've written for kraken2 is great and really helps clarify things.